src/conf/domain_conf.c

```diff
@@ -19897,19 +19901,69 @@ virDomainDefParseXML(xmlXPathContextPtr ctxt,
                                                        ctxt,
                                                        nodes[i],
                                                        flags);
+
         if (!chr)
             return NULL;
 
-        if (chr->target.port == -1) {
+        def->serials[def->nserials++] = chr;
+
+        // Giving precedence to the isa-serial device since
+        // only limited ports can be used for such devices.
+        if (chr->targetType == VIR_DOMAIN_CHR_SERIAL_TARGET_MODEL_ISA_SERIAL) {
+            // Taking the isa serial deivces to start of the array.
+            for (j = def->nserials; j > isa_serial_count; j--)
+                def->serials[j] = def->serials[j-1];
+            def->serials[isa_serial_count++] = chr;
+        }
+
+        // Maintaining the buffer for first max_available_isa_serial_ports unused ports.
+        if (chr->target.port != -1 && chr->target.port <= max_available_isa_serial_ports) {
```

Comment:
Do not you want to give error if it is more than max_available_isa_serial_ports. and max_available_isa_serial_ports should be as some macro.

Reply:
doing it in #19954.

Reply:
yeah I wanted to make it a macro or a gflag but was not sure about the proper location. Will update it.